alacritty/src/renderer/rect.rs

```diff
@@ -0,0 +1,259 @@
+use std::mem::size_of;
+use std::ptr;
+
+use alacritty_terminal::term::SizeInfo;
+
+use crate::gl;
+use crate::gl::types::*;
+use crate::renderer::rects::RenderRect;
+use crate::renderer::{create_program, create_shader, Error, ShaderCreationError};
+
+/// Maxmimum number of rect vertices per batch. Limited by 16-bit integer indices.
+const MAX_VERTICES: usize = u16::max_value() as usize + 1;
+
+/// Shader sources for rect rendering program.
+static RECT_SHADER_F_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.f.glsl");
+static RECT_SHADER_V_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.v.glsl");
+static RECT_SHADER_F: &str = include_str!("../../res/rect.f.glsl");
+static RECT_SHADER_V: &str = include_str!("../../res/rect.v.glsl");
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Rgba {
+    r: u8,
+    g: u8,
+    b: u8,
+    a: u8,
+}
+
+/// Vertex to store solid-color rectangle data.
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Vertex {
+    // Normalized screen coordinates.
+    // TODO these can certainly be i16.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+/// Structure to store and group together rect-related vertices data and GL objects.
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects. VAO stores attribute and ebo bindings.
+    vao: GLuint,
+    vbo: GLuint,
+    ebo: GLuint,
+
+    program: RectShaderProgram,
+
+    // Accumulated vertices data. Allocated space reused between frames.
+    vertices: Vec<Vertex>,
+}
+
+impl RectRenderer {
+    /// Update program when doing live-shader-reload.
+    pub fn set_program(&mut self, program: RectShaderProgram) {
+        self.program = program;
+    }
+
+    pub fn new() -> Result<Self, Error> {
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let mut ebo: GLuint = 0;
+        let program = RectShaderProgram::new()?;
+
+        // Pre-generate the entire index buffer for max possible rects.
+        let mut indices = Vec::<u16>::new();
+        let max_rects = MAX_VERTICES / 4;
+        indices.reserve(max_rects * 6);
+        for index in 0..max_rects {
+            let index = index as u16 * 4;
+            indices.push(index);
+            indices.push(index + 1);
+            indices.push(index + 2);
+            indices.push(index + 2);
+            indices.push(index + 3);
+            indices.push(index + 1);
+        }
```

Comment:
Because this method needs index data, the data is constant, and generating it on-demand requires (a) bookkeeping, (b) reuploading this data to GL.

Alternatively, we could get rid of indices completely. Just generate 6 vertices per rect, vertex shader is trivial anyway. This is ~20% less memory efficient, but also allows to render all quads at once regardless of any limits.

There are more complex alternatives which involve baking rect data into textures, using gl_VertexID (non-GLES-friendly), etc.